lib/backends/webgl/glsl-coordinate-lib.ts

```diff
@@ -54,6 +71,1072 @@ export class CoordsGlslLib extends GlslLib {
+    const rankDiff = outRank - inRank;
+    let coordsSnippet: string;
+    const fields = ['x', 'y', 'z', 'w', 'u', 'v'];
+
+    if (inRank === 0) {
+      coordsSnippet = '';
+    } else if (outRank < 2 && broadcastDims.length >= 1) {
+      coordsSnippet = 'coords = 0;';
+    } else {
+      coordsSnippet = broadcastDims.map(d => `coords.${fields[d + rankDiff]} = 0;`).join('\n');
+    }
+    let unpackedCoordsSnippet = '';
+    if (outRank < 2 && inRank > 0) {
+      unpackedCoordsSnippet = 'coords';
+    } else {
+      unpackedCoordsSnippet = inputLayout.unpackedShape.map((s, i) => `coords.${fields[i + rankDiff]}`).join(', ');
+    }
+    const source = `
+        float ${funcName}() {
+          ${type} coords = getOutputCoords();
+          ${coordsSnippet}
+          return ${texFuncSnippet}(${unpackedCoordsSnippet});
+        }
+      `;
+    return new GlslLibRoutine(source);
+  }
+
+  /**
+   * Constructing snippets for packed operations.
+   */
+  protected getPackedSamplerFromInput(funcName: string, name: string, inputLayout: TextureLayout): GlslLibRoutine {
+    switch (inputLayout.unpackedShape.length) {
+      case 0:
+        return this.getPackedSamplerScalar(funcName, name);
+      case 1:
+        return this.getPackedSampler1D(funcName, name, inputLayout);
+      case 2:
+        return this.getPackedSampler2D(funcName, name, inputLayout);
+      case 3:
+        return this.getPackedSampler3D(funcName, name, inputLayout);
+      default:
+        return this.getPackedSamplerND(funcName, name, inputLayout);
+    }
+  }
+
+  /**
+   * Constructing snippets for unpacked operations.
+   */
+  protected getUnpackedSamplerFromInput(funcName: string, name: string, inputLayout: TextureLayout): GlslLibRoutine {
+    const shape = inputLayout.unpackedShape;
+    switch (shape.length) {
+      case 0:
+        return this.getSamplerScalar(funcName, name, inputLayout);
+      case 1:
+        return this.getSampler1D(funcName, name, inputLayout);
+      case 2:
+        return this.getSampler2D(funcName, name, inputLayout);
+      case 3:
+        return this.getSampler3D(funcName, name, inputLayout);
+      case 4:
+        return this.getSampler4D(funcName, name, inputLayout);
+      case 5:
+        return this.getSampler5D(funcName, name, inputLayout);
+      default:
+        // TODO support more dimensionalities
+        throw new Error(`Unsupported dimension ${shape.length}-D`);
+    }
+  }
+
+  /**
+   * Packed scalar snippet.
+   */
+  protected getPackedSamplerScalar(funcName: string, name: string): GlslLibRoutine {
+    const glsl = getGlsl(this.context.glContext.version);
+    const source = `
+          vec4 ${funcName}() {
+            return ${glsl.texture2D}(${name}, halfCR);
+          }
+        `;
+    return new GlslLibRoutine(source);
+  }
+
+  /**
+   * Unpacked scalar snippet.
+   */
+  protected getSamplerScalar(funcName: string, name: string, inputLayout: TextureLayout): GlslLibRoutine {
```

Comment:
is there any reason why getSamplerXXX() functions are not named as getUnpackedSamplerXXX()?

Reply:
I think it should be getUnpackedSamplerXXX() as it is clearer what this function applies to and more consistent with the getPackedSamplerXXX() ones. Thanks for pointing out. Updated.

---
In reply to: [586195327](https://github.com/microsoft/onnxjs/pull/266#discussion_r586195327) [](ancestors = 586195327)